src/main/scala/zio/nio/channels/DatagramChannel.scala

```diff
@@ -0,0 +1,48 @@
+package zio.nio.channels
+
+import java.io.IOException
+import java.nio.channels.{ DatagramChannel => JDatagramChannel }
+
+import zio.nio.{ ByteBuffer, SocketAddress }
+import zio.{ IO, Managed, UIO }
+
+final class DatagramChannel private[channels] (override protected[channels] val channel: JDatagramChannel)
+    extends GatheringByteChannel
+    with ScatteringByteChannel {
+
+  def bind(local: SocketAddress): IO[IOException, DatagramChannel] =
+    IO.effect(new DatagramChannel(channel.bind(local.jSocketAddress))).refineToOrDie[IOException]
+
+  def connect(remote: SocketAddress): IO[IOException, DatagramChannel] =
+    IO.effect(new DatagramChannel(channel.connect(remote.jSocketAddress))).refineToOrDie[IOException]
+
+  def socket(): UIO[DatagramSocket] =
+    UIO.succeed(new DatagramSocket(channel.socket()))
+
+  def receive(dst: ByteBuffer): IO[IOException, SocketAddress] =
+    IO.effect(new SocketAddress(channel.receive(dst.byteBuffer))).refineToOrDie[IOException]
+
+  def send(src: ByteBuffer, target: SocketAddress): IO[IOException, Int] =
+    IO.effect(channel.send(src.byteBuffer, target.jSocketAddress)).refineToOrDie[IOException]
+
+  def read(dst: ByteBuffer): IO[IOException, Int] =
+    IO.effect(channel.read(dst.byteBuffer)).refineToOrDie[IOException]
+
+  def write(src: ByteBuffer): IO[IOException, Int] =
+    IO.effect(channel.write(src.byteBuffer)).refineToOrDie[IOException]
+}
+
+object DatagramChannel {
+
+  def apply(): Managed[Exception, DatagramChannel] = {
+    val open = IO
+      .effect(JDatagramChannel.open())
+      .refineToOrDie[Exception]
+      .map(new DatagramChannel(_))
+
+    Managed.make(open)(_.close.orDie)
+  }
+
+  def open(): IO[IOException, DatagramChannel] =
```

Comment:
I'm not sure having non-`ZManaged` instantiation is actually useful. If we really need manual acquire/release, I think it can be done by converting the `ZManaged` to a `Reservation` (I think, correct me if I'm wrong).

Reply:
I agree so I removed it for now. If it will be useful in the future it can always be added later.